onnxruntime/core/providers/webgpu/webgpu_context.cc

```diff
@@ -0,0 +1,677 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <memory>
+#include <cmath>
+
+#include "dawn/dawn_proc.h"
+#if !defined(USE_EXTERNAL_DAWN)
+#include "dawn/native/DawnNative.h"
+#endif
+
+#include "core/common/common.h"
+
+#include "core/providers/webgpu/compute_context.h"
+#include "core/providers/webgpu/webgpu_context.h"
+#include "core/providers/webgpu/buffer_manager.h"
+#include "core/providers/webgpu/webgpu_execution_provider.h"
+#include "core/providers/webgpu/program.h"
+#include "core/providers/webgpu/program_cache_key.h"
+#include "core/providers/webgpu/program_manager.h"
+#include "core/providers/webgpu/string_macros.h"
+
+namespace onnxruntime {
+namespace webgpu {
+
+void WebGpuContext::Initialize(const WebGpuExecutionProviderInfo& webgpu_ep_info, const void* dawn_proc_table) {
+  std::call_once(init_flag_, [this, &webgpu_ep_info, dawn_proc_table]() {
+    // Initialization.Step.1 - Create wgpu::Instance
+    if (instance_ == nullptr) {
+      const DawnProcTable* dawn_procs = reinterpret_cast<const DawnProcTable*>(dawn_proc_table);
+#if !defined(USE_EXTERNAL_DAWN)
+      if (dawn_procs == nullptr) {
+        dawn_procs = &dawn::native::GetProcs();
+      }
+#else
+      ORT_ENFORCE(dawn_procs != nullptr, "DawnProcTable must be provided.");
+#endif
+      dawnProcSetProcs(dawn_procs);
```

Comment:
>dawnProcSetProcs [](http://example.com/codeflow?start=6&length=16)

The init_flag_ is per each WebGpuContext. It means we may call this dawnProcSetProcs multiple times in a process? Is it safe?

Reply:
setting dawn proc table once is enough. Usually it is not necessary to set it multiple times. however per our current design , the proc table is passed in session option. so we are doing it like this way.

To answer the question - Yes it is safe.